### PR TITLE
WIP: jobs: Let schedules jobs know when the job completes.

### PR DESCRIPTION
**Pull request by miretskiy**
If the job was started by the scheduled jobs system,
notify scheduled jobs on job completion.

Release Notes: None

### CODE REVIEW

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52419)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/jobs/jobs.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52419#-ME-A4Q83WDz8rck6HfQ:-ME-A4Q83WDz8rck6HfR:b-t12zy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2c66fbaef8f2f6e498c1a15aa2178267bee8e87/pkg/jobs/jobs.go#L559)):*
> ```Go
> 	ctx context.Context, md *JobMetadata, txn *kv.Txn,
> ) error {
> 	if j.createdBy != nil && j.createdBy.Name == CreatedByScheduledJobs {
> ```

I think this is a very nice-to-have feature and it minimally entangling the jobs code with the schedule execution. We could generalize it a bit further and allow registering for termination notifications instead, but i'm usually inclined to wait for the second user before tying to make a generalized API.

But I know spas also wanted to make sure that anything we add here doesn't tangle jobs up more than it is already so i'd also want his eyes on this.

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52419)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @spaskob)

___
*[pkg/jobs/jobs.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52419#-ME-A4Q83WDz8rck6HfQ:-ME75oa37XZsLnI8fd0W:b-phagle) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2c66fbaef8f2f6e498c1a15aa2178267bee8e87/pkg/jobs/jobs.go#L559)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think this is a very nice-to-have feature and it minimally entangling the jobs code with the schedule execution. We could generalize it a bit further and allow registering for termination notifications instead, but i'm usually inclined to wait for the second user before tying to make a generalized API.

But I know spas also wanted to make sure that anything we add here doesn't tangle jobs up more than it is already so i'd also want his eyes on this.
</blockquote></details>

These changes in isolations seem innocuous but as I mentioned before I am worried that this will be death by thousand cuts making it harder and harder to retire old jobs code.

 For example I see that you are already using `jobs.JobMetadata` in scheduler code which is not ideal since it is meant to be used to group the job metadata values passed to UpdateFn in `update.go`.

In addition, you are checking the field `createdBy` in `Job` object. Both these fields in principle are job metadata which should be read from the job row when `job.Update` is called. It  does not make sense for this function to be a method on `Job` since the value of `createdBy` can go out of sync with the values freshly read in `JobMetaData`.

So to sum up, this check should be done instead inside `NotifyJobTermination` since it all ready accepts `JobMetadata` and this function should be removed.

If we are going to do this right, we need to expand `JobMetadata` beyond the scope of `update.go` and also make it include all the job metadata fields.

It will be nice if this is done in a separate PR adding @ajwerner who had some ideas around adding an ImmutableJob struct which looks awfully similar to `JobMetadata`.

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52419)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/jobs/jobs.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52419#-ME-A4Q83WDz8rck6HfQ:-MENayQp9AVZb8rQ8Cs6:bjqwbuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2c66fbaef8f2f6e498c1a15aa2178267bee8e87/pkg/jobs/jobs.go#L559)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

These changes in isolations seem innocuous but as I mentioned before I am worried that this will be death by thousand cuts making it harder and harder to retire old jobs code.

 For example I see that you are already using `jobs.JobMetadata` in scheduler code which is not ideal since it is meant to be used to group the job metadata values passed to UpdateFn in `update.go`.

In addition, you are checking the field `createdBy` in `Job` object. Both these fields in principle are job metadata which should be read from the job row when `job.Update` is called. It  does not make sense for this function to be a method on `Job` since the value of `createdBy` can go out of sync with the values freshly read in `JobMetaData`.

So to sum up, this check should be done instead inside `NotifyJobTermination` since it all ready accepts `JobMetadata` and this function should be removed.

If we are going to do this right, we need to expand `JobMetadata` beyond the scope of `update.go` and also make it include all the job metadata fields.

It will be nice if this is done in a separate PR adding @ajwerner who had some ideas around adding an ImmutableJob struct which looks awfully similar to `JobMetadata`.
</blockquote></details>

As you know, we have tried to avoid adding created_by fields to the jobs -- and tried pretty hard.
I sympathize with your points here, but do disagree with them.  Jobs API is not an API that is clean or an easy to use one.  The fact that we have so many helper "jobs" (startable job, running job, etc) probably should give a hint that job itself is rather messy.  Combined with the fact that registry creates jobs, but jobs ... make calls back into registry to "start" themselves is just one illustration.

Using jobs framework (and schedules jobs MUST use it) entirely from the "outside" the box does not make system easier or more scalable or less error prone.  It makes it harder to understand and pushes unnecessary complexity to the other parts of the code.

There are also additional structural deficiencies present.  For example, is it that unreasonable for a schedule job to know when it completes successfully?  Perhaps publish some metrics that our users want to monitor (backups are important afterall)? I don't think it's unreasonable, but it's also not possible with the resumer interface which only allows OnFailOrCancel.

Another interesting thing is: as a user of the jobs framework "outside" the box, it it safe to make a call back into the jobs framework (e.g. to get jobs current status), while inside resumer callback (OnFailOrCancel)?  Is the callback running under transaction?  Can I start another txn while inside onfailorcance? Again, it's not about the answer to this particular question.  I kinda think it is safe right now, but the moment the code is perturbed, we will have production issues with deadlocks (txn intents) because API doesn't say anything about  the safety of reading jobs table while inside a callback. 

Wrt to the function not belonging to the jobs... Scheduled jobs are called this way because they are *jobs*.  They are not some arbitrary things.  Perhaps we should add a ScheduledJob implementation inside jobs, just like there are StartableJob and the planned ImmutableJob.

At any rate, at this point, this was a WIP/POC to show how straight forward one possible notification implementation might be.  In my book: simple is better.

I will certainly move things to resume/fail callback and drop successful termination notification (though, again, we might actually need it!).  But I do think that in *this* case, this code does belong here, and does make it simpler and, crucially, more correct, and more error prone wrt to the changes that may happen to the jobs implementation itself.

@dt? @ajwerner ?
